cpp/src/arrow/util/bit-util.cc

```diff
@@ -34,13 +44,82 @@ Status BitUtil::BytesToBits(
     const std::vector<uint8_t>& bytes, std::shared_ptr<Buffer>* out) {
   int bit_length = BitUtil::BytesForBits(bytes.size());
 
-  auto buffer = std::make_shared<PoolBuffer>();
-  RETURN_NOT_OK(buffer->Resize(bit_length));
+  std::shared_ptr<MutableBuffer> buffer;
+  RETURN_NOT_OK(AllocateBuffer(default_memory_pool(), bit_length, &buffer));
+
   memset(buffer->mutable_data(), 0, bit_length);
   BytesToBits(bytes, buffer->mutable_data());
 
   *out = buffer;
   return Status::OK();
 }
 
+int64_t CountSetBits(const uint8_t* data, int64_t bit_offset, int64_t length) {
+  constexpr int64_t pop_len = sizeof(uint64_t) * 8;
+
+  int64_t count = 0;
+
+  // The first bit offset where we can use a 64-bit wide hardware popcount
+  const int64_t fast_count_start = BitUtil::RoundUp(bit_offset, pop_len);
+
+  // The number of bits until fast_count_start
+  const int64_t initial_bits = std::min(length, fast_count_start - bit_offset);
+  for (int64_t i = bit_offset; i < bit_offset + initial_bits; ++i) {
```

Comment:
Using the macros defined here https://github.com/apache/parquet-cpp/blob/master/src/parquet/util/bit-util.h#L35 should be a bit faster (or rather more understandable for the compiler).